### PR TITLE
[GOBBLIN-2096] Retry Transient SQL Exception for MALA 

### PR DESCRIPTION
**Pull request by umustafi**
Dear Gobblin maintainers,

Please accept this PR. I understand that it will not be reviewed until I have checked off all the steps below!

### JIRA
- [X ] My PR addresses the following [Gobblin JIRA](https://issues.apache.org/jira/browse/GOBBLIN/) issues and references them in the PR title. For example, "[GOBBLIN-XXX] My Gobblin PR"
    - https://issues.apache.org/jira/browse/GOBBLIN-2096

### Description
- [X] Here are some details about my PR, including screenshots (if applicable):
We're seeing a `MySQLTransactionRollbackException`: Deadlock found when trying to get lock when calling `MysqlMultiActiveLeaseArbiter. attemptLeaseIfNewRow() `in the rare case when two hosts check the lease arbiter table at the same time, see there's no entry matching the current primary key and try to insert it at the exact same time (both wait for lock on the same row). Normally, there's slight discrepancy that results in an `IntegrityConstraintViolationException` if one insert succeeds before the other OR  one finishes the write early and the other sees the entry when reading the `getResult` of the arbiter table. This PR adds retries with exponential backoff for transient SQL exceptions to the insert statement. Note, it's not needed for other cases when the INSERT/UPDATE statements are conditional on a read value. 

### Tests
- [X] My PR adds the following unit tests __OR__ does not need testing for this extremely good reason:
Exponential backoff is tested by other functions, no other major updates here 

### Commits
- [X] My commits all reference JIRA issues in their subject lines, and I have squashed multiple commits if they address the same issue. In addition, my commits follow the guidelines from "[How to write a good git commit message](http://chris.beams.io/posts/git-commit/)":
    1. Subject is separated from body by a blank line
    2. Subject is limited to 50 characters
    3. Subject does not end with a period
    4. Subject uses the imperative mood ("add", not "adding")
    5. Body wraps at 72 characters
    6. Body explains "what" and "why", not "how"

### CODE REVIEW

File: gobblin-service/src/main/java/org/apache/gobblin/service/modules/orchestration/MysqlMultiActiveLeaseArbiter.java [hunk 1]
```diff
@@ -254,7 +256,8 @@ public LeaseAttemptStatus tryAcquireLease(DagActionStore.DagActionLeaseObject da
         log.debug("tryAcquireLease for [{}, is; {}, eventTimestamp: {}] - CASE 1: no existing row for this dag action,"
             + " then go ahead and insert", dagActionLeaseObject.getDagAction(),
             dagActionLeaseObject.isReminder() ? "reminder" : "original", dagActionLeaseObject.getEventTimeMillis());
-        int numRowsUpdated = attemptLeaseIfNewRow(dagActionLeaseObject.getDagAction());
+        int numRowsUpdated = attemptLeaseIfNewRow(dagActionLeaseObject.getDagAction(),
+            ExponentialBackoff.builder().maxRetries(3).initialDelay(10L).build());
```
Comment (Will-Lo) [inline gobblin-service/src/main/java/org/apache/gobblin/service/modules/orchestration/MysqlMultiActiveLeaseArbiter.java:260]: Can we set the defaults as static variables and include them with units? If it's just 10ms delay it wouldn't be much more than a jitter

**Review by Will-Lo — CHANGES_REQUESTED** [review]

File: gobblin-service/src/main/java/org/apache/gobblin/service/modules/orchestration/MysqlMultiActiveLeaseArbiter.java (hunk 1 above)
Comment (arjun4084346) [inline gobblin-service/src/main/java/org/apache/gobblin/service/modules/orchestration/MysqlMultiActiveLeaseArbiter.java:260]: create it one time statically outside the method.

Comment (umustafi) [reply gobblin-service/src/main/java/org/apache/gobblin/service/modules/orchestration/MysqlMultiActiveLeaseArbiter.java:260]: this object is re-used so it needs to be recreated every time it's called

File: gobblin-service/src/test/java/org/apache/gobblin/service/modules/orchestration/MysqlMultiActiveLeaseArbiterTest.java [hunk 2]
```diff
@@ -201,9 +206,11 @@ Tests attemptLeaseIfNewRow() method to ensure a new row is inserted if no row ma
   @Test
   public void testAcquireLeaseIfNewRow() throws IOException {
     // Inserting the first time should update 1 row
-    Assert.assertEquals(mysqlMultiActiveLeaseArbiter.attemptLeaseIfNewRow(resumeDagAction), 1);
+    Assert.assertEquals(mysqlMultiActiveLeaseArbiter.attemptLeaseIfNewRow(resumeDagAction,
+        ExponentialBackoff.builder().maxRetries(3).initialDelay(10L).build()), 1);
```
Comment (arjun4084346) [inline gobblin-service/src/test/java/org/apache/gobblin/service/modules/orchestration/MysqlMultiActiveLeaseArbiterTest.java:210]: same here, though it is less impactful in tests

File: gobblin-service/src/main/java/org/apache/gobblin/service/modules/orchestration/MysqlMultiActiveLeaseArbiter.java (hunk 1 above)
Comment (umustafi) [reply gobblin-service/src/main/java/org/apache/gobblin/service/modules/orchestration/MysqlMultiActiveLeaseArbiter.java:260]: I set defaults for the static variable but I do actually want it to function as jitter for retrying a deadlock exception so I added some randomness to the initial delay. There may be other exceptions we want larger delay for, what do you think as (30,60ms) for the value?

File: gobblin-service/src/main/java/org/apache/gobblin/service/modules/orchestration/MysqlMultiActiveLeaseArbiter.java [hunk 3]
```diff
@@ -171,6 +173,8 @@ public class MysqlMultiActiveLeaseArbiter implements MultiActiveLeaseArbiter {
   // Complete lease acquisition if values have not changed since lease was acquired
   protected static final String CONDITIONALLY_COMPLETE_LEASE_STATEMENT = "UPDATE %s SET "
       + "event_timestamp=event_timestamp, lease_acquisition_timestamp = NULL " + WHERE_CLAUSE_TO_MATCH_ROW;
+  protected static final int MAX_RETRIES = 3;
+  protected static final long INITIAL_DELAY_FOR_RETRY_MILLIS = 30L;
   private static final ThreadLocal<Calendar> UTC_CAL =
```
Comment (arjun4084346) [inline gobblin-service/src/main/java/org/apache/gobblin/service/modules/orchestration/MysqlMultiActiveLeaseArbiter.java:178]:
```
private static ExponentialBackoff exponentialBackOff = ExponentialBackoff.builder().maxRetries(MAX_RETRIES)
                .initialDelay(INITIAL_DELAY_FOR_RETRY_MILLIS + (long) Math.random() * INITIAL_DELAY_FOR_RETRY_MILLIS)
                .build();
```

File: gobblin-service/src/main/java/org/apache/gobblin/service/modules/orchestration/MysqlMultiActiveLeaseArbiter.java [hunk 4]
```diff
@@ -420,15 +427,25 @@ protected GetEventInfoResult createGetInfoResult(ResultSet resultSet) throws IOE
...
           completeInsertPreparedStatement(insertStatement, dagAction);
           try {
             return insertStatement.executeUpdate();
-          } catch (SQLIntegrityConstraintViolationException e) {
+          } catch (SQLTransientException e) {
+            try {
+              if (exponentialBackoff.awaitNextRetryIfAvailable()) {
+                return attemptLeaseIfNewRow(dagAction, exponentialBackoff);
+              }
+            } catch (InterruptedException | IOException e2) {
```
Comment (arjun4084346) [inline gobblin-service/src/main/java/org/apache/gobblin/service/modules/orchestration/MysqlMultiActiveLeaseArbiter.java:443]: IOException is not required, is it?

File: gobblin-service/src/main/java/org/apache/gobblin/service/modules/orchestration/MysqlMultiActiveLeaseArbiter.java [hunk 5]
```diff
@@ -420,15 +428,25 @@ protected GetEventInfoResult createGetInfoResult(ResultSet resultSet) throws IOE
...
+              throw new IOException(e2);
+            }
+            return 0;
```
Comment (arjun4084346) [inline gobblin-service/src/main/java/org/apache/gobblin/service/modules/orchestration/MysqlMultiActiveLeaseArbiter.java:447]: this line will reach when there is a continuous SQLTransientException that exponential backoff could not help with. in that case should we thrown the original exception as it is?

Comment (umustafi) [reply gobblin-service/src/main/java/org/apache/gobblin/service/modules/orchestration/MysqlMultiActiveLeaseArbiter.java:447]: Oh good call, doing that

File: gobblin-service/src/main/java/org/apache/gobblin/service/modules/orchestration/MysqlMultiActiveLeaseArbiter.java (hunk 1 above)
Comment (umustafi) [reply gobblin-service/src/main/java/org/apache/gobblin/service/modules/orchestration/MysqlMultiActiveLeaseArbiter.java]: increased to be in range 20,220

Comment (Will-Lo) [reply gobblin-service/src/main/java/org/apache/gobblin/service/modules/orchestration/MysqlMultiActiveLeaseArbiter.java]: Sounds good

**Review by Will-Lo — APPROVED** [review]

**Pull request merged**